es-vue-base/src/lib-components/EsFileInput.vue

```diff
@@ -0,0 +1,266 @@
+    },
+    watch: {
+        uploadUrls: {
+            handler(newUrls) {
+                if (newUrls.length > 0 && newUrls.length === this.files.length) this.upload();
+            },
+            deep: true,
+        },
+        files: {
+            handler(newFiles) {
+                if (newFiles.length > 0 && newFiles.length === this.uploadUrls.length) this.upload();
+            },
+            deep: true,
+        },
+        pickedItems(newVal) {
+            // The user has selected files from the file picker. We have to filter out any files that exceed the
+            // maxFileSizeMb prop.
+            this.files = this.filterLargeFiles(newVal);
+            if (this.files.length > 0) {
+                this.$emit('readyToUpload', this.files.length);
+                this.readFilesIntoUrl(newVal);
+            }
+        },
+    },
+    methods: {
+        filterLargeFiles(files) {
+            // Takes an array of files, and filters out any files that exceed the maxFileSizeMb prop. Emits a
+            // fileSizeError event for each file that exceeds the maxFileSizeMb prop. The File API uses bytes, so
+            // we have to convert the maxFileSizeMb prop to bytes.
+            const maxFileSizeBytes = this.maxFileSizeMb * 1000000;
+            return files.filter((file) => {
+                if (file.size > maxFileSizeBytes) {
+                    this.$emit('fileSizeError', file.name);
+                    return false;
+                }
+                return true;
+            });
+        },
+        readFilesIntoUrl(files) {
+            this.fileUrls = [];
+            files
+                .forEach((file) => {
+                    const fileReader = new FileReader();
+                    fileReader.onload = () => {
+                        // TODO: If pdf/docx, we might not want to read the file...
+                        this.$emit('fileDataRead', {
+                            name: file.name, type: file.type, size: file.size, data: fileReader.result,
+                        });
+                    };
+                    fileReader.readAsDataURL(file);
+                });
+        },
+        onDrop(event) {
+            // The user has dropped files onto the component. We have to apply the same logic as if they had
+            // selected the files from the file picker which limits the file types to the ones specified in the
+            // fileTypes prop.
+            this.active = false;
+            this.files = [];
+
+            // Use DataTransferItemList interface to access the file(s)
+            const invavlidItems = [...event.dataTransfer.items].filter((item) => item.kind !== 'file');
+
+            const dataTransfersAsFiles = [...event.dataTransfer.items]
+                .filter((item) => item.kind === 'file')
+                .map((item) => item.getAsFile());
+
+            const validFileTypes = dataTransfersAsFiles
+                .filter((file) => this.fileTypes.includes(file.type));
```

Comment:
So this does not seem good enough to me on second glance. It appears that the File API just [checks the extension](https://developer.mozilla.org/en-US/docs/Web/API/File/type) which can be spoofed. I am going to revise this to check the mimetype [with magic numbers](https://medium.com/the-everyday-developer/detect-file-mime-type-using-magic-numbers-and-javascript-16bc513d4e1e)

Reply:
And done

Reply:
Very clever.